### PR TITLE
Refactor widgets code

### PR DESCRIPTION
**Pull request by dbkr**
Split logic out of setWidget & re-use it when we add jitsi widgets.

Also fix the detection of existing jitsi widgets and make that use
getRoomWidgets() too.

Add error message when a jitsi call is already in process.

Also merges the two different widget utils files into one

### CODE REVIEW

File: src/WidgetUtils.js [hunk 1]
```diff
@@ -254,4 +253,83 @@ export default class WidgetUtils {
...
+     */
+    static getUserWidgets() {
+        const client = MatrixClientPeg.get();
+        if (!client) {
+            throw new Error('User not logged in');
+        }
+        const userWidgets = client.getAccountData('m.widgets');
+        let userWidgetContent = {};
+        if (userWidgets && userWidgets.getContent()) {
+          userWidgetContent = userWidgets.getContent();
```
Comment (lukebarnard1) [inline src/WidgetUtils.js]: This feels.. linted.

File: src/WidgetUtils.js [hunk 2]
```diff
@@ -254,4 +253,83 @@ export default class WidgetUtils {
...
+        }
+        return userWidgetContent;
+    }
+
+    /**
+     * Get user specific widgets (not linked to a specific room) as an array
+     * @return {[object]} Array containing current / active user widgets
+     */
+    static getUserWidgetsArray() {
+      return Object.values(WidgetUtils.getUserWidgets());
```
Comment (lukebarnard1) [inline src/WidgetUtils.js]: More dodgy indentation.

File: src/CallHandler.js [hunk 3]
```diff
@@ -61,6 +61,8 @@ import Matrix from 'matrix-js-sdk';
 import dis from './dispatcher';
 import { showUnknownDeviceDialogForCalls } from './cryptodevices';
 import SettingsStore from "./settings/SettingsStore";
+import WidgetUtils from './WidgetUtils';
+import { getRoomWidgets } from './utils/widgets';
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/CallHandler.js]: Yes, so was I :)

File: src/WidgetUtils.js [hunk 4]
```diff
@@ -190,4 +192,66 @@ export default class WidgetUtils {
...
+            name: widgetName,
+            data: widgetData,
+        };
+
+        const client = MatrixClientPeg.get();
+        const userWidgets = Widgets.getUserWidgets();
+
+        // Delete existing widget with ID
+        try {
+            delete userWidgets[widgetId];
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/WidgetUtils.js]: This is one for @rxl881

File: src/WidgetUtils.js [hunk 5]
```diff
@@ -190,4 +192,66 @@ export default class WidgetUtils {
...
+        } catch (e) {
+            console.error(`$widgetId is non-configurable`);
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/WidgetUtils.js]: @rxl881 again

File: src/WidgetUtils.js [hunk 6]
```diff
@@ -190,4 +192,66 @@ export default class WidgetUtils {
...
+        }
+
+        // Add new widget / update
+        if (widgetUrl !== null) {
+            userWidgets[widgetId] = {
+                content: content,
+                sender: client.getUserId(),
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/WidgetUtils.js]: defer to @rxl881 again

File: src/WidgetUtils.js [hunk 7]
```diff
@@ -254,4 +253,83 @@ export default class WidgetUtils {
...
         });
     }
+
+    /**
+     * Get all widgets (user and room) for the current user
+     * @param  {object} room The room to get widgets for
+     * @return {[object]} Array containing current / active room and user widget state events
+     */
+    static getWidgets(room) {
+        const widgets = getRoomWidgets(room);
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/WidgetUtils.js]: Yes, but this function is removed now as it wasn't used.

File: src/WidgetUtils.js [hunk 8]
```diff
@@ -254,4 +253,83 @@ export default class WidgetUtils {
...
+        return widgets;
+    }
+
+    /**
+     * Get room specific widgets
+     * @param  {object} room The room to get widgets force
+     * @return {[object]} Array containing current / active room widgets
+     */
+    static getRoomWidgets(room) {
+        const appsStateEvents = room.currentState.getStateEvents('im.vector.modular.widgets');
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/WidgetUtils.js]: Don't believe so - it's just advertised as a property of the room object.

File: src/WidgetUtils.js (hunk 1 above)
Comment (dbkr) [reply src/WidgetUtils.js]: mm, also incorrectly indented

File: src/components/views/elements/AppTile.js [hunk 9]
```diff
@@ -319,14 +319,15 @@ export default class AppTile extends React.Component {
...
+
+                        WidgetUtils.setRoomWidget(
                             this.props.id,
-                        ).then(() => {
-                            return WidgetUtils.waitForRoomWidget(this.props.id, this.props.room.roomId, false);
-                        }).catch((e) => {
+                            null,
+                            null,
+                            null,
+                            null,
```
Comment (lukebarnard1) [inline src/components/views/elements/AppTile.js]: That is a lot of `null`s. Would it be better to wrap that in `deleteRoomWidget`?

**Pull request merged**